### ISSUE TITLE
Metaprogramming in Nim #1 Introduction and Generic Procedures

### ISSUE THREAD

**Issue by Kiloneie** [nim-lang/edutainment#4]
In this video i will introduce the concept of metaprogramming to the viewer, it's 3 metaprogramming constructs/features with the most basic template example for copy pasta and then move to some equally easy to understand Generic Procedures(procs).

Chapters/sub-chapters:

- Chapter: What is metaprogramming ?
-  sub chapter: Metaprogramming use cases
- Chapter: Generic Procedures
- sub-chapter: constraining generics
- sub-chapter: style note(which construct/feature to use in order for readability and to not waste time when you don't need a macro)

**Comment by Kiloneie** [comment]
Near finished video, awaiting for feedback, link: https://youtu.be/gkEb5azqkoc

**Comment by HugoGranstrom** [comment]
Nice video :) One thing I hung up on though, is that you repeatedly said that you can't have generics of different types, which is true when you only use one generic. But it sounds like you say that it isn't possible at all, which could be giving the wrong impression of generics and making them look less useful than they are. Idk how much detail to go into in an intro video to meta-programming, but I would rephrase those parts to at least not be so dismissive of the fact that it is possible. 

Eg: 
```nim
proc hello[T, U](a: T, b: U) =
	...
```

**Comment by Kiloneie** [comment]
That's a good point, i will put that into the video, some other video will then go into finer detail.

**Comment by Kiloneie** [comment]
Here is a version 2 of this video's script: https://play.nim-lang.org/#ix=3BC5

**Comment by Araq** [comment]
Here are my corrections: https://play.nim-lang.org/#ix=3Cnx 
Please diff these two and take care of the details.

**Comment by Kiloneie** [comment]
Here is the video with all the corrections applied and some extras i added afterwards(mainly the SDL part being demonstrated instead of just code written for it - template part).

Link: https://youtu.be/2EWLE-IFgGM

**Comment by Kiloneie** [comment]
Video complete, closing.
